src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/DependencyGraph.java

```diff
@@ -0,0 +1,202 @@
+package at.ac.tuwien.kr.alpha.common.depgraph;
+
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import at.ac.tuwien.kr.alpha.common.Predicate;
+import at.ac.tuwien.kr.alpha.common.atoms.Atom;
+import at.ac.tuwien.kr.alpha.common.atoms.Literal;
+import at.ac.tuwien.kr.alpha.grounder.NonGroundRule;
+
+// TODO ensure immutability
+public class DependencyGraph {
+
+	private static final Logger LOGGER = LoggerFactory.getLogger(DependencyGraph.class);
+
+	private static final String CONSTRAINT_PREDICATE_FORMAT = "[constr_%d]";
+
+	/**
+	 * Maps nodes to outgoing edges of that node NOTE: Doing the value as List<Edge> rather than Map<Integer,Boolean> as one node may have positive and negative
+	 * edges to another.
+	 */
+	private Map<Node, List<Edge>> nodes = new HashMap<>();
+
+	private Map<Predicate, Node> nodesByPredicate = new HashMap<>();
+
+	/**
+	 * The transposed graph structure, i.e. the same set of nodes, but all edges reversed - needed for analysis of strongly connected components.
+	 */
+	private Map<Node, List<Edge>> transposedNodes;
+
+	private Map<Integer, List<Node>> stronglyConnectedComponents;
+
+	private int constraintNumber;
+
+	public DependencyGraph() {
+		// doing this since checkstyle doesn't let me initialize it directly to 0
+		this.constraintNumber = 0;
+	}
+
+	public static DependencyGraph buildDependencyGraph(Map<Integer, NonGroundRule> nonGroundRules) {
+		DependencyGraph retVal = new DependencyGraph();
+		NonGroundRule tmpRule;
+		Node tmpHeadNode;
+		for (Map.Entry<Integer, NonGroundRule> entry : nonGroundRules.entrySet()) {
+			tmpRule = entry.getValue();
+			LOGGER.debug("Processing rule: {}", tmpRule);
+			tmpHeadNode = retVal.handleRuleHead(tmpRule);
+			for (Literal l : entry.getValue().getBodyLiterals()) {
+				LOGGER.trace("Processing rule body literal: {}", l);
+				if (l.getAtom().isBuiltin()) {
+					LOGGER.trace("Ignoring builtin atom in literal {}", l);
+					continue;
+				}
+				retVal.handleRuleBodyLiteral(tmpHeadNode, l);
+			}
+		}
+		retVal.findStronglyConnectedComponents();
+		return retVal;
+	}
+
+	public Node getNodeForPredicate(Predicate p) {
+		return this.nodesByPredicate.get(p);
+	}
+
+	private Node handleRuleHead(NonGroundRule rule) {
+		LOGGER.trace("Processing head of rule: {}", rule);
+		Node retVal;
+		Predicate pred;
+		if (rule.isConstraint()) {
+			pred = this.generateConstraintDummyPredicate();
```

Comment:
There are many seemingly superfluous `this.` around, please remove them wherever possible.